tests/test_modeling_fsmt.py

```diff
@@ -0,0 +1,509 @@
+        )
+        self.assertEqual(
+            output_new.shape,
+            (new_src_vocab_size, config.d_model),
+            msg=f"output {input_new.shape}, {(new_src_vocab_size, config.d_model)}",
+        )
+        self.assertTrue(torch.eq(input_new, output_new).all(), msg=f"{input_new}, {output_new}")
+
+
+def _assert_tensors_equal(a, b, atol=1e-12, prefix=""):
+    """If tensors not close, or a and b arent both tensors, raise a nice Assertion error."""
+    if a is None and b is None:
+        return True
+    try:
+        if torch.allclose(a, b, atol=atol):
+            return True
+        raise
+    except Exception:
+        msg = "{} != {}".format(a, b)
+        if prefix:
+            msg = prefix + ": " + msg
+        raise AssertionError(msg)
+
+
+def _long_tensor(tok_lst):
+    return torch.tensor(tok_lst, dtype=torch.long, device=torch_device)
+
+
+TOLERANCE = 1e-4
+
+
+@require_torch
+class FSMTModelIntegrationTests(unittest.TestCase):
+    @cached_property
+    def default_tokenizer(self):
+        return FSMTTokenizer.from_pretrained("stas/fsmt-wmt19-ru-en")
+
+    @slow
+    def test_inference_no_head(self):
+        tokenizer = self.default_tokenizer
+        model = FSMTModel.from_pretrained("stas/fsmt-wmt19-ru-en").to(torch_device)
+
+        src_text = "My friend computer will translate this for me"
+        input_ids = tokenizer([src_text], return_tensors="pt")["input_ids"]
+        input_ids = _long_tensor(input_ids)
+        inputs_dict = prepare_fsmt_inputs_dict(model.config, input_ids)
+        with torch.no_grad():
+            output = model(**inputs_dict)[0]
+        expected_shape = torch.Size((1, model.config.decoder_attention_heads, model.config.tgt_vocab_size))
+        self.assertEqual(output.shape, expected_shape)
+        expected_slice = torch.tensor(
+            [[-3.1850, -3.1849, 2.9694], [-4.0242, -4.0242, 0.2494], [-3.4442, -3.4443, 0.3315]], device=torch_device
+        )
+        print(output[:, :3, :3])
+        self.assertTrue(torch.allclose(output[:, :3, :3], expected_slice, atol=TOLERANCE))
+
+        # XXX: the rest of the tests were moved to tests/test_tokenization_bart.py - port from there into tokenization tests
+
+    @slow
+    def test_translation(self):
+        text = {
+            "en": "Machine learning is great, isn't it?",
+            "ru": "Машинное обучение - это здорово, не так ли?",
+            "de": "Maschinelles Lernen ist großartig, oder?",
+        }
+
+        pairs = [
+            ["en", "ru"],
+            ["ru", "en"],
+            ["en", "de"],
+            ["de", "en"],
+        ]
+
+        for src, tgt in pairs:
+            print(f"Testing {src} -> {tgt}")
+            mname = f"stas/fsmt-wmt19-{src}-{tgt}"
+
+            src_sentence = text[src]
+            tgt_sentence = text[tgt]
+
+            tokenizer = FSMTTokenizer.from_pretrained(mname)
+            model = FSMTForConditionalGeneration.from_pretrained(mname)
```

Comment:
send this `to(torch_device)`, and maybe fp16.

Reply:
I took this suggestion as all of them at once. Let me know if you were proposing to test each one of them separately (i.e. 3 distinct tests).